src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java

```diff
@@ -0,0 +1,116 @@
+package org.jabref.logic.citationstyle;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.jabref.model.entry.BibEntry;
+
+import de.undercouch.citeproc.CSL;
+import de.undercouch.citeproc.ItemDataProvider;
+import de.undercouch.citeproc.ListItemDataProvider;
+import de.undercouch.citeproc.bibtex.BibTeXConverter;
+import de.undercouch.citeproc.csl.CSLItemData;
+import de.undercouch.citeproc.output.Bibliography;
+import org.jbibtex.BibTeXEntry;
+import org.jbibtex.DigitStringValue;
+import org.jbibtex.Key;
+
+/**
+ * Provides an adapter class to CSL. It holds a CSL instance under the hood that is only recreated when
+ * the style changes.
+ *
+ * @apiNote The first call to {@link #makeBibliography} is expensive since the
+ * CSL instance will be created. As long as the style stays the same, we can reuse this instance. On style-change, the
+ * engine is re-instantiated. Therefore, the use-case of this class is many calls to {@link #makeBibliography} with the
+ * same style. Changing the output format is cheap.
+ * @implNote The main function {@link #makeBibliography} will enforce
+ * synchronized calling. The main CSL engine under the hood is not thread-safe. Since this class is usually called from
+ * a SwingWorker, the only other option would be to create several CSL instances which is wasting a lot of resources and very slow.
+ * In the current scheme, {@link #makeBibliography} can be called as usual
+ * SwingWorker task and to the best of my knowledge, concurrent calls will pile up and processed sequentially.
+ */
+public class CSLAdapter {
+
+    private static final BibTeXConverter BIBTEX_CONVERTER = new BibTeXConverter();
+    private static String style;
+    private static CitationStyleOutputFormat format;
+    private static JabRefItemDataProvider dataProvider = new JabRefItemDataProvider();
+    private static CSL cslInstance;
+
+    /**
+     * Creates the bibliography of the provided items. This method needs to run synchronized because the underlying
+     * CSL engine is not thread-safe.
+     */
+    static synchronized List<String> makeBibliography(List<BibEntry> bibEntries, String style, CitationStyleOutputFormat outputFormat) throws IOException {
```

Comment:
As far as I can see by looking at the code, this solution should be thread-safe. You could turn up parallelism by a tiny notch, but I cannot say if this has any tangible benefit. This could be done by not marking the whole method as synchronized, but just the part where the CSL engine is used. To do so, you would add a private lock instance variable, e.g. `private static Object lock = new Object();` and then refactor the method like this:
```
        List<CSLItemData> cslItemData = new ArrayList<>(bibEntries.size());
        for (BibEntry bibEntry : bibEntries) {
            cslItemData.add(bibEntryToCSLItemData(bibEntry));
        }

        synchronized(lock) {
            initialize(style, outputFormat);
            dataProvider.setData(cslItemData);
            cslInstance.registerCitationItems(dataProvider.getIds());
            Bibliography bibliography = cslInstance.makeBibliography();
            return Arrays.asList(bibliography.getEntries());
        }
```
You can refactor the code in this way or not, I leave this up to you. It will only have an effect if there are many concurrent calls to the method.

Reply:
It is an easy fix indeed. I will profile where the bottleneck is when you have many items selected and click on `Copy Preview`. It might be useful to check if the loop itself should be parallelized as well. Copying the preview is too slow for my taste and I'm unsure it really comes from CSL this time. I will check.

Reply:
I profiled this and it's having no significant impact on the performance. After refactoring the creation of the `cslItemData` was absorbed into our data provider and must be run synchronized as well or the data we provide to CSL will get out of synch.